### PR TITLE
 signatures: Add data_end_offset to signature_match() and custom events

### PR DESCRIPTION
**Pull request by awelzel**
This is a replacement/rework for #3908. Add the end offset to the `signature_match()` event as an optional parameter.
This is a bit cumbersome as we also support custom events for signatures, but seems to work out.

```
event signature_match(state: signature_state, msg: string, data: string, data_end_offset: string) {
    print data[:data_end_offset];
}
```

@rsmmr - how does this look/feel to you? The #3908 approach added the offset to `state`, but the offset (particularly the relative one for `data` that is wanted) isn't very meaningful within the `state` by itself.

I'm not super fond of all the `signature_match()` permutations we now support, but maybe that's okay.

### CODE REVIEW

File: NEWS [hunk 1]
```diff
@@ -62,6 +62,11 @@ New Functionality
 * The DNS analyzer was extended to support TKEY RRs (RFC 2390). A corresponding
   ``dns_TKEY`` event was added.
 
+* The ``signature_match()`` and custom signature events receive the relative
+  end offset for the match within data as an optional parameter.
+
+	event signature_match(state: signature_state, msg: string, data: string, data_end_offset: count);
+
```
Comment (rsmmr) [inline NEWS:69]: Before I look more closely, let me double check that I understand the semantics correctly: is `data_end_offset` the offset of the final byte of the match inside *data*, or the offset of the final byte of `data` itself?

Comment (awelzel) [reply NEWS:69]: The former and actually one behind the match. But yes, relative within the current `data` instance.

S for two packets with payloads
```
AAAAA
BXBBB
```
and `/.*X.*/` as signature, data would be populated with `BXBBB` and data_end_offset should be `2`.

The name isn't great. Maybe `data_match_size` instead?

(I'm also going to produce a pcap with above traffic to make sure this is 100% correct :-) )

**Review by rsmmr — APPROVED** [review]
This seems to ok to me. It's a bit messy to get that new information through everywhere, in particular (I agree) with the various event signatures we now support, but not much to do about that. Looking at the end result, I like this more than putting it into `signature_state`. So if we do want to get this offset into scriptland, this is the best I can think of, too.

File: src/RuleAction.cc [hunk 2]
```diff
@@ -27,42 +31,88 @@ RuleActionEvent::RuleActionEvent(const char* arg_msg, const char* event_name) {
     handler = zeek::event_registry->Lookup(event_name);
 
     if ( ! handler ) {
-        reporter->Error("unknown event '%s' specified in rule", event_name);
-        return;
+        // Has it been declared?
+        const auto& id = zeek::id::find(event_name);
+        if ( id->GetType()->Tag() != TYPE_FUNC || id->GetType()->AsFuncType()->Flavor() != FUNC_FLAVOR_EVENT ) {
```
[the remark this answers is not in the document]
Comment (awelzel) [reply src/RuleAction.cc:36]: Yeah, sorry that was a bit of a fly-by. I realized when writing tests that if you only declare a custom event but never implement it we currently fail with an error. I'm happy to revert this and fix at another time.

[the remark this answers is not in the document]
Comment (rsmmr) [reply src/RuleAction.cc:36]: No problem, that's fine, just couldn't mentally connect it. :-)

File: src/RuleMatcher.h [hunk 3]
```diff
@@ -182,12 +184,15 @@ class RuleEndpointState {
...
 
     // The follow tracks which rules for which all patterns have matched,
-    // and in a parallel list the (first instance of the) corresponding
-    // matched text.
+    // in a parallel list the (first instance of the) corresponding
+    // matched text, and in another parallel list the offset of the
+    // end of the last pattern match.
     rule_list matched_by_patterns;
     bstr_list matched_text;
+    match_offset_list matched_text_end_offsets;
```
[the remark this answers is not in the document]
Comment (awelzel) [reply src/RuleMatcher.h]: I'm going to merge this PR as is. I've prepped the rework, but I"m not comfortable merging without another look.

**Pull request merged**